astropy/visualization/wcsaxes/wcsapi.py

```diff
@@ -0,0 +1,254 @@
+
+    def __init__(self, wcs, invert_xy=False):
+
+        super().__init__()
+
+        if wcs.pixel_n_dim != 2:
+            raise ValueError('Only pixel_n_dim==2 is supported')
+
+        self.wcs = wcs
+        self.invert_xy = invert_xy
+
+        self.frame_in = wcsapi_to_celestial_frame(wcs)
+
+    def __eq__(self, other):
+        # TODO: for performance, find a way to return True when they do match
+        return False
+
+    @property
+    def input_dims(self):
+        return self.wcs.world_n_dim
+
+    def transform(self, world):
+
+        # Convert to a list of arrays
+        world = list(world.T)
+
+        if len(world) != self.wcs.world_n_dim:
+            raise ValueError(f"Expected {self.wcs.world_n_dim} world coordinates, got {len(world)} ")
+
+        if len(world[0]) == 0:
+            pixel = np.zeros((0, 2))
+        else:
+            pixel = self.wcs.world_to_pixel_values(*world)
+
+        if self.invert_xy:
+            pixel = pixel[::-1]
+
+        pixel = np.array(pixel).T
+
+        return pixel
+
+    transform_non_affine = transform
+
+    def inverted(self):
+        """
+        Return the inverse of the transform
+        """
+        return WCSPixel2WorldTransform(self.wcs, invert_xy=self.invert_xy)
+
+
+class WCSPixel2WorldTransform(CurvedTransform):
+    """
+    WCS transformation from pixel to world coordinates
+    """
+
+    has_inverse = True
+
+    def __init__(self, wcs, invert_xy=False):
+
+        super().__init__()
+
+        if wcs.pixel_n_dim != 2:
+            raise ValueError('Only pixel_n_dim==2 is supported')
+
+        self.wcs = wcs
+        self.invert_xy = invert_xy
+
+        self.frame_out = wcsapi_to_celestial_frame(wcs)
+
+    def __eq__(self, other):
+        # TODO: for performance, find a way to return True when they do match
+        return False
+
+    @property
+    def output_dims(self):
+        return self.wcs.world_n_dim
+
+    def transform(self, pixel):
+
+        # Convert to a list of arrays
+        pixel = list(pixel.T)
+
+        if len(pixel) != self.wcs.pixel_n_dim:
+            raise ValueError(f"Expected {self.wcs.pixel_n_dim} world coordinates, got {len(pixel)} ")
+
+        if self.invert_xy:
+            pixel = pixel[::-1]
+
+        if len(pixel[0]) == 0:
+            world = np.zeros((0, self.wcs.world_n_dim))
+        else:
+            world = self.wcs.pixel_to_world_values(*pixel)
+
+        # At the moment, one has to manually check that the transformation
+        # round-trips, otherwise it should be considered invalid.
+        pixel_check = self.wcs.pixel_to_world_values(*world)
+        with np.errstate(invalid='ignore'):
+            invalid = np.zeros(len(pixel[0]), dtype=bool)
+            for ipix in range(len(pixel)):
+                invalid |= np.abs(pixel_check[ipix] - pixel[ipix]) > 1.
+            for iwrl in range(len(world)):
+                world[iwrl][invalid] = np.nan
+
+        world = np.array(world).T
+
+        return world
```

Comment:
This can return >2D right? Is this supposed to do that and the downstream code works out which ones it wants?

Reply:
In principle yes but in practice I think you wouldn't want to do that as Matplotlib doesn't know what to do with 3-d output.

Reply:
But I mean despite pixel dim being two here, we could end up with more than two world coords? So don't we need to use the slice information to reduce dimensionality to two here?

Reply:
Yes that's correct, we do support >2 world outputs here, sorry!

Reply:
Ping @astrofrog @Cadair 
Has the above issue been solved?

Reply:
I think it works properly, I don't think there was an issue?

Reply:
it looks like this PR has changed the length of `WCSAxes.coords` for >2D WCSes. i.e this https://github.com/sunpy/sunpy/pull/3283/commits/8d42d7b0b559911a114a2a71565291d559b7a239 commit